src/hotspot/share/nmt/nmtTreap.hpp

```diff
@@ -0,0 +1,389 @@
+      if (maximum_depth_found < head.depth) {
+        maximum_depth_found = head.depth;
+      }
+      if (head.parent_prio < head.n->_priority) {
+        return false;
+      }
+      to_visit.push({head.depth + 1, head.n->_priority, head.n->left()});
+      to_visit.push({head.depth + 1, head.n->_priority, head.n->right()});
+    }
+    if (maximum_depth_found > (int)expected_maximum_depth) {
+      return false;
+    }
+    // Visit everything in order, see that the key ordering is monotonically increasing.
+    TreapNode* last_seen = nullptr;
+    bool failed = false;
+    this->visit_in_order([&](TreapNode* node) {
+      if (last_seen == nullptr) {
+        last_seen = node;
+        return;
+      }
+      int c = COMPARATOR::cmp(last_seen->key(), node->key());
+      if (c > 0) {
+        failed = false;
+      }
+      last_seen = node;
+    });
+    return !failed;
+  }
+#endif // ASSERT
+
+public:
+  Treap(uint64_t seed = static_cast<uint64_t>(os::random()))
+  : _root(nullptr),
+  _prng_seed(seed)
+  DEBUG_ONLY(COMMA _node_count(0)) {
+  }
+
+  ~Treap() {
+    this->remove_all();
+  }
+
+  void upsert(const K& k, const V& v) {
+    assert(verify_self(), "invariant");
+
+    TreapNode* found = find(_root, k);
+    if (found != nullptr) {
+      // Already exists, update value.
+      found->_value = v;
+      return;
+    }
+    DEBUG_ONLY(_node_count++;)
+    // Doesn't exist, make node
+    void* node_place = ALLOCATOR::allocate(sizeof(TreapNode));
+    uint64_t prio = prng_next();
+    TreapNode* node = new (node_place) TreapNode(k, v, prio);
+
+    // (LEQ_k, GT_k)
+    node_pair split_up = split(this->_root, k);
+    // merge(merge(LEQ_k, EQ_k), GT_k)
+    this->_root = merge(merge(split_up.left, node), split_up.right);
+  }
+
+  void remove(const K& k) {
+    assert(verify_self(), "invariant");
+
+    // (LEQ_k, GT_k)
+    node_pair fst_split = split(this->_root, k, LEQ);
+    // (LT_k, GEQ_k) == (LT_k, EQ_k) since it's from LEQ_k and keys are unique.
+    node_pair snd_split = split(fst_split.left, k, LT);
+
+    if (snd_split.right != nullptr) {
+      // The key k existed, we delete it.
+      DEBUG_ONLY(_node_count--;)
+      ALLOCATOR::free(snd_split.right);
+    }
+    // Merge together everything
+    this->_root = merge(snd_split.left, fst_split.right);
+  }
+
+  // Delete all nodes.
+  void remove_all() {
+    DEBUG_ONLY(_node_count = 0;)
+    GrowableArrayCHeap<TreapNode*, mtNMT> to_delete;
+    to_delete.push(this->_root);
+
+    while (!to_delete.is_empty()) {
+      TreapNode* head = to_delete.pop();
+      if (head == nullptr) continue;
+      to_delete.push(head->_left);
+      to_delete.push(head->_right);
+      ALLOCATOR::free(head);
+    }
+  }
+
+  TreapNode* closest_geq(const K& key) {
+    // Need to go "left-ward" for EQ node, so do a leq search first.
+    TreapNode* leqB = closest_leq(key);
+    if (leqB != nullptr && leqB->key() == key) {
+      return leqB;
+    }
```

Comment:
I don't get this, why is this leq search needed? And if its needed, is that not redundant to the code below that compares for key equality?